Inbox redesign to use channel folders.

amanagr opened the pull request:
Inbox doesn't update right when channel folders are renamed or updated for channels. I can open a separate PR to do that.

alya commented:
We should probably have tooltips:

- Expand channel folder / Collapse channel folder
- Expand channel / Collapse channel
- Go to list of topics / Go to channel feed

timabbott commented on web/styles/inbox.css:
```diff
@@ -18,6 +18,14 @@
+            border-radius: 3px;
+            color: var(--color-inbox-row-unread-count);
+            font-variant: all-small-caps;
+            font-weight: 400;
```
Something seems to have gone wrong with the sizing of unread count pills; they should be consistent in width for a given number of digits. Note how the `1`s are misaligned with the rest:

![image](https://github.com/user-attachments/assets/6e57c308-c64b-4a88-a473-998c3cf8d591)

amanagr replied:
Fixed in this update.

<img width="134" height="494" alt="image" src="https://github.com/user-attachments/assets/043da4e5-9313-48af-92d2-c2b7c5978761" />

timabbott commented:
Some observations from manual testing:
- I think the unread counts on folders might look better with pills around them, and a less faint font color. Right now, at least in dark theme, they just fade into the background:
![image](https://github.com/user-attachments/assets/d3c854c2-3356-4567-80b1-9e558d3a89f6)
- If you're in a filtered state (like followed topics) or otherwise only have unreads in "OTHER CHANNELS", somehow it feels weird to have the only channels heading be "OTHER CHANNELS". I'm not sure I have a good answer for that.
![image](https://github.com/user-attachments/assets/f837ef08-64ce-445b-ae77-aef6daa9c6c9)

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -597,12 +638,35 @@ function rerender_stream_inbox_header_if_needed(
+    new_folder_data: ChannelFolderContext,
+): void {
+    for (const property of channel_folder_context_properties) {
+        if (new_folder_data[property] !== old_folder_data[property]) {
```
This loop is a bit weird; I feel like a single `if (_.any(...))` clause comparing the fields  ... or just an `_.isEqual` would be more natural?

amanagr replied:
I can open a separate PR for this. Currently it follows the pattern of other `rerender_foo_if_needed` functions.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -766,6 +826,27 @@ function get_sorted_row_dict<T extends DirectMessageContext | TopicContext>(
+function sort_channel_folders(): void {
+    const sorted_channel_folders = [...channel_folders_dict.values()].sort((a, b) => {
+        // Sort OTHER_CHANNELS_FOLDER_ID last, then by name with PINNED_CHANNEL_FOLDER_ID first.
+        if (a.id === OTHER_CHANNELS_FOLDER_ID) {
```
I feel like you might be able to just use a normal sorting function after mapping `OTHER_CHANNELS_FOLDER_ID` to `Infinity` for sorting, but maybe that has its own issues involving `Infinity-Infinity`. This seems fine.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -766,6 +826,27 @@ function get_sorted_row_dict<T extends DirectMessageContext | TopicContext>(
+        if (b.id === PINNED_CHANNEL_FOLDER_ID) {
+            return 1;
+        }
+        return util.strcmp(a.name, b.name);
```
I think it'd be best to have this call a new `channel_folders.compare_folders()` function that is this `strcmp` thing, so that this code just works if/when we change the sort key on them. Minor.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -821,6 +903,83 @@ function reset_data(): {
     topics_dict = get_sorted_stream_topic_dict();
     const is_dms_collapsed = collapsed_containers.has("inbox-dm-header");
 
+    for (const [, stream_data] of streams_dict) {
```
`stream_data` seems like a bad variable, given we have a module by the same name. Maybe just `stream_context`, to match the type, would be better?

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -821,6 +903,83 @@ function reset_data(): {
+                    channel_folder.has_unread_mention || stream_data.mention_in_unread;
+            }
+        }
+    }
```
I think you can probably deduplicate this a bunch with a couple functions -- one to make a `folder_context` object (probably a better name than `channel_folder`, since the latter should be reserved for an API object), and a `update_unread_counts(folder_context, channel_context)` for the second part of this repeated pattern.

amanagr replied:
Moved the entire `if/else` clause to a new function.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -1195,9 +1345,18 @@ function is_list_focused(): boolean {
 }
 
 function get_all_rows(): JQuery {
```
Maybe add a docstring to this function saying what at a high level it's supposed to be.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -1626,7 +1785,42 @@ export function change_focused_element(input_key: string): boolean {
+    let previous_folder_id;
+    for (const [folder_id, channel_folder] of channel_folders_dict) {
+        if (channel_folders.has(folder_id)) {
+            const $channel_folder = render_inbox_folder_with_channels({
```
`$folder_row_html` seems like a better name -- we generally want `_html` suffixes on rendered HTML. And again `folder_context` for the loop variable.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -1740,6 +1964,64 @@ export function update(): void {
+            name = $t({defaultMessage: "OTHER CHANNELS"});
+        } else {
+            name = channel_folders.get_channel_folder_name_from_id(folder_id);
+        }
```
I feel like a function to do this would be useful both here and in the loop for https://github.com/zulip/zulip/pull/35152/commits/7b8a6ef27b5a1145322670f3d3ac901df6a6367a#r2196185448.

amanagr replied:
Yeah, done.

timabbott commented on web/src/unread.ts:
```diff
@@ -1042,6 +1042,10 @@ export function num_unread_mentions_in_dms(): number {
 
+export function num_unread_mentions_in_topics(): number {
+    return unread_mention_topics.size;
+}
```
This looks unused?

amanagr replied:
Yeah, removed.

alya commented:
Yeah, I was clicking on counters on muted topics and on channels, not on folders and DM headers, and the state wasn't live-updating properly.

amanagr commented:
> We should probably have tooltips:
> 
> * Expand channel folder / Collapse channel folder
> * Expand channel / Collapse channel
> * Go to list of topics / Go to channel feed

I can do this as a followup.

> Is keyboard nav expected to be working at this point? Pressing Enter always collapses/expands a channel, regardless of which button appears to be selected.

Yes, it is working as expected for me. Can you tell me what you are doing step by step so I can reproduce.



> Yeah, I was clicking on counters on muted topics and on channels, not on folders and DM headers, and the state wasn't live-updating properly.

Fixed it. Thanks for catching.


> If you're in a filtered state (like followed topics) or otherwise only have unreads in "OTHER CHANNELS", somehow it feels weird to have the only channels heading be "OTHER CHANNELS". I'm not sure I have a good answer for that.

Fixed.

> I think the unread counts on folders might look better with pills around them, and a less faint font color. Right now, at least in dark theme, they just fade into the background:

I think we should deploy it to get better feedback on how to fix this?

amanagr commented on web/src/inbox_ui.ts:
```diff
@@ -821,6 +903,83 @@ function reset_data(): {
+        const other_channels_folder = channel_folders_dict.get(OTHER_CHANNELS_FOLDER_ID);
+        if (other_channels_folder !== undefined) {
+            other_channels_folder.name = $t({defaultMessage: "CHANNELS"});
+        }
```
No since else part is handled by `rerender_channel_folder_header_if_needed`.

timabbott replied:
Hmm, OK. Feels a bit fragile, but maybe it's fine.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -645,26 +710,21 @@ function format_topic(
-    topic_dict: Map<string, {topic_count: number; latest_msg_id: number}>,
-): boolean {
-    const stream_key = get_stream_key(stream_id);
-    update_stream_data(stream_id, stream_key, topic_dict);
```
OK I agree that is correct.

amanagr commented:
> We should probably have tooltips:
> 
> * Expand channel folder / Collapse channel folder
> * Expand channel / Collapse channel
> * Go to list of topics / Go to channel feed

<img width="1378" height="315" alt="image" src="https://github.com/user-attachments/assets/31449e88-0a13-4da0-aad0-1a5c58e19106" />

amanagr commented:
I will push a fix to perfectly align channel name and topic name, I see they got a litttle misaligned after the recent changes.

amanagr commented:
> Let's use the chevron colors from Figma; there are are couple of lighter colors on the channel chevrons (different for collapsed vs. hovered channels, it looks like). We can do something similar in dark theme if there's no specific design.

Should we try to match colors in left sidebar here?

alya commented:
OK, for some reason "Go to list of topics" in the channel menu doesn't work for me if I select it with the keyboard. Other options work fine with the keyboard, and that option works fine with the mouse.

alya commented:
That's all I've got for this round of manual testing! No blockers to deploying on CZO from my perspective.

amanagr commented:
> OK, for some reason "Go to list of topics" in the channel menu doesn't work for me if I select it with the keyboard. Other options work fine with the keyboard, and that option works fine with the mouse.

Fixed

amanagr commented:
There are 2 pending issues here:

* When the inbox view is really tall, it seems like when the view rerenders (like due to a message coming in), I was getting scrolled to the top. So I was having a hard time looking at a channel group that's a few screens down.  
 - Debugging this but don't know why this is happening.

* Let's use the chevron colors from Figma; there are are couple of lighter colors on the channel chevrons (different for collapsed vs. hovered channels, it looks like). We can do something similar in dark theme if there's no specific design.
 - Waiting for dark more colors to implement this.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -1422,6 +1422,10 @@ export function get_focused_row_message(): {message?: Message | undefined} & (
+        if (Number.isNaN(stream_id)) {
+            // This is a channel folder header.
+            return {};
+        }
```
This check should probably be a `data-folder-id` check sort of thing.

timabbott commented:
I'm not sure if this is new, but there's a scroll position flicker I'm seeing when navigating back to inbox using the `Back` button after starting from a scrolled-down position:

[Screencast from 2025-07-18 16-16-16.webm](https://github.com/user-attachments/assets/fa5001ed-7ecc-481a-b168-1e11d9570e19)

timabbott commented on web/src/channel_folders.ts:
```diff
@@ -39,3 +40,9 @@ export function get_channel_folders(include_archived = false): ChannelFolder[] {
+export function get_channel_folder_name_from_id(folder_id: number): string {
+    const channel_folder = channel_folder_by_id_dict.get(folder_id);
+    assert(channel_folder !== undefined);
+    return channel_folder.name;
```
We should just offer a proper getter here:

``` diff
diff --git a/web/src/channel_folders.ts b/web/src/channel_folders.ts
index 6afaf49402..bd4a803333 100644
--- a/web/src/channel_folders.ts
+++ b/web/src/channel_folders.ts
@@ -41,8 +41,8 @@ export function is_valid_folder_id(folder_id: number): boolean {
     return channel_folder_by_id_dict.has(folder_id);
 }
 
-export function get_channel_folder_name_from_id(folder_id: number): string {
+export function get_channel_folder_by_id(folder_id: number): ChannelFolder {
     const channel_folder = channel_folder_by_id_dict.get(folder_id);
     assert(channel_folder !== undefined);
-    return channel_folder.name;
+    return channel_folder;
 }
diff --git a/web/src/inbox_ui.ts b/web/src/inbox_ui.ts
index a249477e8c..0bd3b53b4d 100644
--- a/web/src/inbox_ui.ts
+++ b/web/src/inbox_ui.ts
@@ -852,7 +852,7 @@ function get_folder_name_from_id(folder_id: number): string {
         return $t({defaultMessage: "OTHER CHANNELS"});
     }
 
-    return channel_folders.get_channel_folder_name_from_id(folder_id);
+    return channel_folders.get_channel_folder_by_id(folder_id).name;
 }
 
 function update_channel_folder_data(channel_context: StreamContext): void {
```

timabbott replied:
@evykassirer FYI, since I think you add more or less exactly this function in the left sidebar PR, and this will be a hopefully trivial merge conflict for you to resolve.

timabbott commented on web/src/inbox_ui.ts:
```diff
@@ -830,6 +958,20 @@ function reset_data(): {
     };
 }
 
+function is_only_other_channels_folder_visible(): boolean {
```
I guess `is_other_channels_only_visible_folder` may be better, went with that.

timabbott commented:
OK, with the above changes, I've merged this, thanks @amanagr! TODO list is:
- The two issues noted in https://github.com/zulip/zulip/pull/35152#issuecomment-3091231697
- Refactoring this check: https://github.com/zulip/zulip/pull/35152#pullrequestreview-3034756349
- The chevron color issue noted here: https://github.com/zulip/zulip/pull/35152#issuecomment-3083581612. I'll note I no longer experience the scroll-position-jumping that I'd been seeing on chat.zulip.org, so I'm assuming you did something to fix it (but don't know what that was).
